### DISCUSSION TITLE
Shipping odin for macOS without requiring the llvm installation from homebrew

### DISCUSSION THREAD

**Discussion by hasenj** [odin-lang/Odin#1742] [Proposals]
I've always been [concerned][0] about how the official distribution of odin for macOS depends on having LLVM installed on the system in a specific location.

[0]: https://github.com/odin-lang/Odin/issues/1013

If you download the odin binary for macos from the github releases page you get a ~2.7MB file that requires llvm to be installed at `/usr/local/opt/llvm@11/`, or at least that's what the error message is telling me:

```
dyld[12062]: Library not loaded: /usr/local/opt/llvm@11/lib/libLLVM-C.dylib
  Referenced from: /Users/hasenj/odin/odin
  Reason: tried: '/usr/local/opt/llvm@11/lib/libLLVM-C.dylib' (no such file), '/usr/local/lib/libLLVM-C.dylib' (no such file), '/usr/lib/libLLVM-C.dylib' (no such file)
Abort trap: 6
```

I suspect the path is this way because of how homebrew sets things up.

I removed the llvm installed from brew on my system (long story) and instead downloaded the [official llvm release][1] (which at this time was [version 14][2]) and it is basically just a folder. So I copied the folder to a specific location on my system (`~/programs/llvm`) and added the `bin` directory from there to my PATH environment variable.

[1]: https://github.com/llvm/llvm-project/releases
[2]: https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.1/clang+llvm-14.0.1-x86_64-apple-darwin.tar.xz

Now, I wanted to compile odin using _that_ installation of llvm, and I wanted odin to be statically linked so that I can delete the llvm folder (or rename it) and still have the odin compiler work.

I struggled for a while but it worked after I made the following changes to the build script:

```diff
❯ git diff
diff --git a/build_odin.sh b/build_odin.sh
index aef3f283..0fc44f97 100755
--- a/build_odin.sh
+++ b/build_odin.sh
@@ -4,7 +4,7 @@ set -eu
 GIT_SHA=$(git rev-parse --short HEAD)
 DISABLED_WARNINGS="-Wno-switch -Wno-macro-redefined -Wno-unused-value"
 LDFLAGS="-pthread -lm -lstdc++"
-CFLAGS="-std=c++14 -DGIT_SHA=\"$GIT_SHA\""
+CFLAGS="-std=c++14 -DGIT_SHA=\"$GIT_SHA\" -isysroot $(xcrun --show-sdk-path)"
 CFLAGS="$CFLAGS -DODIN_VERSION_RAW=\"dev-$(date +"%Y-%m")\""
 CC=clang
 OS=$(uname)
@@ -38,7 +38,7 @@ config_darwin() {
 
        LDFLAGS="$LDFLAGS -liconv -ldl"
        CFLAGS="$CFLAGS $($LLVM_CONFIG --cxxflags --ldflags)"
-       LDFLAGS="$LDFLAGS -lLLVM-C"
+       LDFLAGS="$LDFLAGS $($LLVM_CONFIG --libs --system-libs)"
 }
 
 config_freebsd() {
```

The resulting binary was roughly 90MB, which was a good sign because it means llvm is statically linked into it.

I renamed the llvm folder on my system so that if the compiler tries to look for it, it won't find it. Then I used the compiler, and it worked!

I successfully created a build of the compiler that does not have any runtime dependencies beside those present on the system.

Basically I needed to add this ` -isysroot $(xcrun --show-sdk-path)` to the compiler flags because the official llvm distribution for macos can't seem to find the system sdk on its own.

Replacing `-lLLVM-C` with `$($LLVM_CONFIG --libs --system-libs)` probably has to do with how homebrew sets things up. I suppose the `LLVM-C` is the clurpit causing the llvm libraries to be dynamically instead of statically linked. I'm not sure where it comes from, but it's not included in the official llvm distribution that I downloaded.

It seems the build script for macOS not only requires llvm, but assumes that llvm is installed via homebrew. Because, why else would it try to link against `LLVM-C` instead of asking llvm-config to provide the library names?

Running `dyld_info odin` against the version I compiled with the above changes shows it's only dynamically linked to system libraries:

    /usr/lib/libSystem.B.dylib
    @rpath/libc++.1.dylib
    /usr/lib/libiconv.2.dylib
    /usr/lib/libz.1.dylib
    /usr/lib/libncurses.5.4.dylib
    /usr/lib/libxml2.2.dylib

Where as running the same command on the official odin distributed compiler shows:

    /usr/lib/libSystem.B.dylib
    /usr/local/opt/llvm@11/lib/libc++.1.dylib
    /usr/lib/libiconv.2.dylib
    /usr/local/opt/llvm@11/lib/libLLVM-C.dylib

I suspect some system libraries (libz, libxml, libcurses) are not listed because they are dependencies of LLVM-C, and the `dyld_info` tool only prints the direct dependencies.

This means that if someone downloads the official llvm distribution and copies it to the path required by the odin compiler, it still will not work, because there's no `libLLVM-C.dylib` in it.

In other words, not only does odin currently require llvm to be installed, it also requires homebrew to be installed, and for llvm to be installed via homebrew.

**Comment by sumofat** [comment]
I agree there is something not always right when I installed odin on my mac's.
